File: src/main/java/com/imyyq/mvvm/binding/viewadapter/recyclerview/LayoutManagers.java
package com.imyyq.mvvm.binding.viewadapter.recyclerview;

import androidx.annotation.IntDef;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.StaggeredGridLayoutManager;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

public class LayoutManagers {
    protected LayoutManagers() {
    }

    public interface LayoutManagerFactory {
        RecyclerView.LayoutManager create(RecyclerView recyclerView);
    }

    /**
     * A {@link LinearLayoutManager}.
     */
    public static LayoutManagerFactory linear() {
        return recyclerView -> new LinearLayoutManager(recyclerView.getContext());
    }

    /**
     * A {@link LinearLayoutManager} with the given orientation and reverseLayout.
     */
    public static LayoutManagerFactory linear(@Orientation final int orientation, final boolean reverseLayout) {
        return recyclerView -> new LinearLayoutManager(recyclerView.getContext(), orientation, reverseLayout);
    }

    /**
     * A {@link GridLayoutManager} with the given spanCount.
     */
    public static LayoutManagerFactory grid(final int spanCount) {
        return recyclerView -> new GridLayoutManager(recyclerView.getContext(), spanCount);
    }

    public static LayoutManagerFactory grid(final int spanCount, final GridLayoutManager.SpanSizeLookup lookup) {
        return recyclerView -> {
            GridLayoutManager manager = new GridLayoutManager(recyclerView.getContext(), spanCount);
            manager.setSpanSizeLookup(lookup);
            return manager;
        };
    }

    /**
     * A {@link GridLayoutManager} with the given spanCount, orientation and reverseLayout.
     **/
    public static LayoutManagerFactory grid(final int spanCount, @Orientation final int orientation, final boolean reverseLayout) {
        return recyclerView -> new GridLayoutManager(recyclerView.getContext(), spanCount, orientation, reverseLayout);
    }

    /**
     * A {@link StaggeredGridLayoutManager} with the given spanCount and orientation.
     */
    public static LayoutManagerFactory staggeredGrid(final int spanCount, @Orientation final int orientation) {
        return recyclerView -> new StaggeredGridLayoutManager(spanCount, orientation);
    }

    @IntDef({LinearLayoutManager.HORIZONTAL, LinearLayoutManager.VERTICAL})
    @Retention(RetentionPolicy.SOURCE)
    public @interface Orientation {
    }
}

File: src/main/java/com/imyyq/mvvm/binding/viewadapter/recyclerview/LayoutManagers.kt
package com.imyyq.mvvm.binding.viewadapter.recyclerview

import androidx.annotation.IntDef
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.GridLayoutManager.SpanSizeLookup
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.StaggeredGridLayoutManager
import com.imyyq.mvvm.binding.command.BindingFunction

/**
 * RecyclerView 的布局管理器类，提供便捷的方法
 */
object LayoutManagers {
    /**
     * 垂直线性布局，从上往下
     */
    @JvmStatic
    fun verticalLinear(): BindingFunction<RecyclerView, RecyclerView.LayoutManager> {
        return BindingFunction { LinearLayoutManager(it.context) }
    }

    /**
     * 线性布局，可控制方向
     */
    @JvmStatic
    fun linear(
        @Orientation orientation: Int,
        reverseLayout: Boolean
    ): BindingFunction<RecyclerView, RecyclerView.LayoutManager> {
        return BindingFunction {
            LinearLayoutManager(
                it.context,
                orientation,
                reverseLayout
            )
        }
    }

    /**
     * 垂直网格布局
     */
    @JvmStatic
    fun verticalGrid(spanCount: Int): BindingFunction<RecyclerView, RecyclerView.LayoutManager> {
        return BindingFunction {
            GridLayoutManager(
                it.context,
                spanCount
            )
        }
    }

    /**
     * 垂直的网格布局
     *
     * @param spanCount 多少列
     * @param lookup    控制每个布局占多少列
     */
    @JvmStatic
    fun verticalGrid(
        spanCount: Int,
        lookup: SpanSizeLookup?
    ): BindingFunction<RecyclerView, RecyclerView.LayoutManager> {
        return BindingFunction {
            val manager = GridLayoutManager(it.context, spanCount)
            manager.spanSizeLookup = lookup
            manager
        }
    }

    /**
     * 网格布局，可控制方向
     */
    @JvmStatic
    fun grid(
        spanCount: Int,
        @Orientation orientation: Int,
        reverseLayout: Boolean
    ): BindingFunction<RecyclerView, RecyclerView.LayoutManager> {
        return BindingFunction {
            GridLayoutManager(
                it.context,
                spanCount,
                orientation,
                reverseLayout
            )
        }
    }

    /**
     * 交错网格布局
     */
    @JvmStatic
    fun staggeredGrid(
        spanCount: Int,
        @Orientation orientation: Int
    ): BindingFunction<RecyclerView, RecyclerView.LayoutManager> {
        return BindingFunction { it: RecyclerView? ->
            StaggeredGridLayoutManager(
                spanCount,
                orientation
            )
        }
    }

    @IntDef(LinearLayoutManager.HORIZONTAL, LinearLayoutManager.VERTICAL)
    @Retention(AnnotationRetention.SOURCE)
    annotation class Orientation
}